### DISCUSSION TITLE
Android FluidSynth audio output speakers / earphones

### DISCUSSION THREAD

**Discussion by nahumbudin** [FluidSynth/fluidsynth#931] [Q&A] [answered]
Hi,

I'm running an native  android fluidsynth (2.18) application I have wrote on an Android 10 realme 6 device.
It works perfect.

However, if I connect an earphones (or a cable to an audio amp) to the device, the sound stops. 
I need to restart the application while the earphones plug is connected, to get the audio out.
The same if I now unplug the earphones - I need to restart the app to get the audio back from the phone speakers. 

Any idea why? 
Any fix/workaround?

Thanks,

Nahum Budin.

**Comment by derselbst** [comment]
@atsushieno FYI

**Comment by atsushieno** [comment, marked as answer]
Hi,

When I implemented Oboe / OpenSLES driver, I wasn't aware of such audio device disconnection matters, hence there is no special care about it. I'm not sure how Fluidsynth itself deals with disconnected devices - is it taken care on other audio drivers? @derselbst 

This is how Oboe detects device disconnection.
https://github.com/google/oboe/blob/master/docs/FullGuide.md#disconnected-audio-stream

Maybe we can implement something. That means, there is no workaround so far.

**Reply by derselbst** [reply]
Sry, I'm not aware of any special treatment of device disconnection in any of our existing audio drivers. It would probably just end in some error.

**Reply by atsushieno** [reply]
Actually using `audio.driver=opensles` settings might be a workaround, IF adjusting sample rate is non-issue. https://github.com/google/oboe/blob/master/docs/notes/disconnect.md#tech-note-disconnected-streams-and-plugin-issues

> When Oboe is using OpenSL ES, and a headset is plugged in or out, then OpenSL ES will automatically switch between devices. This is convenient but can cause problems because the new device may have different burst sizes and different latency.

**Reply by nahumbudin** [reply]
Hey guys,

As you have suggested ,I have changed

 fluid_settings_setstr(settings, "audio.driver", "oboe"); 

to

 fluid_settings_setstr(settings, "audio.driver", "opensles");

and it looks like it is solving the issue: I can now switch earphones in and out, and audio resumes to the proper device:
speaker/earphones, as expected.

Thanks a lot for you support.

Nahum.

**Reply by atsushieno** [reply]
https://github.com/FluidSynth/fluidsynth/pull/940 - implements auto reconnection (with additional settings key).
